evm/chains/header.py

```diff
@@ -0,0 +1,156 @@
+from abc import ABCMeta, abstractmethod
+from typing import Dict, Any, Tuple, Type  # noqa: F401
+
+from eth_typing import (
+    BlockNumber,
+    Hash32,
+)
+
+from evm.db.backends.base import BaseDB
+from evm.db.header import (  # noqa: F401
+    BaseHeaderDB,
+    HeaderDB,
+)
+from evm.rlp.headers import BlockHeader
+from evm.utils.datatypes import (
+    Configurable,
+)
+from evm.vm.base import BaseVM  # noqa: F401
+
+
+class BaseHeaderChain(Configurable, metaclass=ABCMeta):
+    _basedb = None  # type: BaseDB
+
+    _headerdb_class = None  # type: Type[BaseHeaderDB]
+    _headerdb = None  # type: BaseHeaderDB
+
+    header = None  # type: BlockHeader
+    network_id = None  # type: int
+    vm_configuration = None  # type: Tuple[Tuple[int, Type[BaseVM]], ...]
+
+    @abstractmethod
+    def __init__(self, basedb: BaseDB, header: BlockHeader) -> None:
+        raise NotImplementedError("Chain classes must implement this method")
+
+    #
+    # Chain Initialization API
+    #
+    @classmethod
+    @abstractmethod
+    def from_genesis_header(cls,
+                            basedb: BaseDB,
+                            genesis_header: BlockHeader) -> 'BaseHeaderChain':
+        raise NotImplementedError("Chain classes must implement this method")
+
+    #
+    # Helpers
+    #
+    @classmethod
+    @abstractmethod
+    def get_headerdb_class(cls):
+        raise NotImplementedError("Chain classes must implement this method")
+
+    #
+    # Header API
+    #
+    @abstractmethod
+    def get_canonical_block_header_by_number(self, block_number: BlockNumber) -> BlockHeader:
+        raise NotImplementedError("Chain classes must implement this method")
+
+    @abstractmethod
+    def get_canonical_head(self) -> BlockHeader:
+        raise NotImplementedError("Chain classes must implement this method")
+
+    @abstractmethod
+    def get_block_header_by_hash(self, block_hash: Hash32) -> BlockHeader:
+        raise NotImplementedError("Chain classes must implement this method")
+
+    @abstractmethod
+    def import_header(self, header: BlockHeader) -> Tuple[BlockHeader, ...]:
+        raise NotImplementedError("Chain classes must implement this method")
+
+
+class HeaderChain(BaseHeaderChain):
```

Comment:
I don't understand why this class is reimplementing all these headerdb methods as pass-throughs. Can it just have the couple of extra methods it needs, and expose the headerdb? Just seeing if we can cut down on total code weight.

Reply:
I'm not sold this is the right way to go either, however, I was trying to not expose the database directly.  Going to think on this a bit and see if I can talk myself in-to/out-of it.  

I'm thinking about this in the same way that the `Chain` proxies some of the `VM` APIs.  I forsee (with casper) having the `VM` specify it's own class of `HeaderDB` at some point when the format of individual headers changes.

Reply:
Still seems like we can cover that case by exposing the header db as a `BaseHeaderDB` from both `HeaderChain` implementations.

`Chain` & `VM` seems a bit more reasonable to duplicate the methods (though I'm not sure that's the right solution either), because the VM is different depending on the block. But `HeaderDB` is always the same for the given `HeaderChain`.

...unless we want to do `HeaderChain` -> `VM` by block number -> `HeaderDB`. But I don't think we need to store headers differently by fork, even if the header format changes, so that's probably unnecessary complexity.